onnxscript/rewriter/ort_fusions/sdpa.py

```diff
@@ -8,24 +8,67 @@
 
 
 class SDPA(pattern.RewriteRuleClassBase):
-    def __init__(self, name: str, *, use_mask: bool, pre_scale: bool, use_mul: bool):
-        super().__init__(name=name, as_function=True)
+    def __init__(
+        self,
+        name: str,
+        *,
+        use_mask: bool,
+        pre_scale: bool,
+        pre_scale_q: bool,
+        use_mul: bool,
+        has_3d_inputs: bool,
+    ):
+        super().__init__(name=name)
         self._use_mask = use_mask
         self._pre_scale = pre_scale
+        # There are some patterns where only the query is scaled before the dot product
+        # and essentially (query * qk_scale) * key is equivalent to (query * key) * qk_scale
+        # TODO: Capture patterns where only the key is scaled before the dot product
+        self._pre_scale_q = pre_scale_q
         self._use_mul = use_mul
+        # Capture patterns where the query is reshaped from 3D to 4D
+        # after scaling has been applied to query.
+        self._has_3d_inputs = has_3d_inputs
         self._scale: float | None = None
 
     def pattern(
-        self, op, query, key_transposed, value, mask, query_scale, key_scale, qk_scale
+        self,
+        op,
+        query,
+        key_transposed,
+        value,
+        mask,
+        query_scale,
+        key_scale,
+        qk_scale,
+        # Shape used for reshaping the query in patterns where query is reshaped
+        # from 3D to 4D and scaling is applied before the reshaping.
+        query_reshape,
     ):
         if self._pre_scale:
             # Some implementations scale the query and key before computing the dot product
             if self._use_mul:
-                query = op.Mul(query, query_scale)
-                key_transposed = op.Mul(key_transposed, key_scale)
+                if self._pre_scale_q:
+                    query = op.Mul(query, qk_scale)
+                else:
+                    query = op.Mul(query, query_scale)
+                    key_transposed = op.Mul(key_transposed, key_scale)
             else:
-                query = op.Div(query, query_scale)
-                key_transposed = op.Div(key_transposed, key_scale)
+                if self._pre_scale_q:
+                    query = op.Div(query, qk_scale)
+                else:
+                    query = op.Div(query, query_scale)
+                    key_transposed = op.Div(key_transposed, key_scale)
+
+        # There might be patterns where the reshape and transpose are done
+        # after the pre-scaling. If the inputs are 3D, we need to reshape them to 4D
+        # and apply the approriate transposes to query.
+        if self._has_3d_inputs and self._pre_scale_q:
```

Comment:
Do these 2 options interact? Doesn't 3D/4D variation affect all (query/key/value)?

Reply:
If all query/key/values are just 3D or just 4D, SDPA should be able to handle them. But this a very specific scenario in whisper, where the Mul for the query happens before the Reshape/Transpose, so we are looking at 3D query but a 4D key/value.

`pre_scale_q` handles the case where only q is scaled however it doesn't care whether query is 3D or 4D, as there could be cases where just 4D query is pre-scaled as well.

Now, `has_3d_inputs` which in retrospect might be confusing as a name, only checks in cases where query is pre-scaled, whether that query was 3d/4d. I think changing the parameter name to `has_3d_query` might be a better idea for now.